File: 10.Math数学/leetcode_2069_Robot.cpp
class Robot {
private:
    int w, h, pos, tot;
public:
    Robot(int width, int height) {
        tot = 2*(width+height)-4;
        w = width;
        h = height;
        pos = 0;
    }
    
    void step(int num) {
        pos = (pos+num-1)%tot+1;
    }
    
    vector<int> getPos() {
        if (pos <= w-1){
            return {pos, 0};
        }
        else if (pos <= w+h-2){
            return {w-1, pos-w+1};
        }
        else if (pos <= 2*w+h-3){
            return {w-1-(pos-w-h+2), h-1};
        }
        return {0, h-1-(pos-2*w-h+3)};
    }
    
    string getDir() {
        if (pos <= w-1){
            return "East";
        }
        else if (pos <= w+h-2){
            return "North";
        }
        else if (pos <= 2*w+h-3){
            return "West";
        }
        return "South";
    }
};

/**
 * Your Robot object will be instantiated and called as such:
 * Robot* obj = new Robot(width, height);
 * obj->step(num);
 * vector<int> param_2 = obj->getPos();
 * string param_3 = obj->getDir();
 */

File: 10.Math数学/leetcode_2069_Robot.py
from typing import List

class Robot:

    def __init__(self, width: int, height: int):
        self.w, self.h = width, height
        self.totLength = 2*(width+height)-4
        self.pos = 0
        self.start = False

    def step(self, num: int) -> None:
        self.start = True
        self.pos = (self.pos+num-1)%self.totLength+1
    
    def _getState(self):
        if self.pos <= self.w-1:
            return [self.pos, 0, "East"]
        if self.pos <= self.w+self.h-2:
            return [self.w-1, self.pos-self.w+1, "North"]
        if self.pos <= 2*self.w+self.h-3:
            return [self.w-(self.pos-self.w-self.h+2)-1, self.h-1, "West"]
        return [0, self.h-(self.pos-2*self.w-self.h+3)-1, "South"]

    def getPos(self) -> List[int]:
        return self._getState()[:2]

    def getDir(self) -> str:
        return self._getState()[2]

if __name__ == "__main__":
    r = Robot(6,3)
    ops = ["step","step","getPos","getDir","step","step","step","getPos","getDir"]
    nums = [[2],[2],[],[],[2],[1],[4],[],[]]
    for i in range(len(ops)):
        numStr = ""
        if len(nums[i]):
            numStr = str(nums[i][0])
        cmdstr = "r."+ops[i]+"("+numStr+")"
        print(eval(cmdstr))
